src/app/fdctl/configure/ethtool-gro.c

```diff
@@ -0,0 +1,184 @@
+  return !err;
+}
+
+static void
+device_read_slaves( const char * device,
+                    char         output[ 4096 ] ) {
+  char path[ PATH_MAX ];
+  FD_TEST( fd_cstr_printf_check( path, PATH_MAX, NULL, "/sys/class/net/%s/bonding/slaves", device ) );
+
+  FILE * fp = fopen( path, "r" );
+  if( FD_UNLIKELY( !fp ) )
+    FD_LOG_ERR(( "error configuring network device, fopen(%s) failed (%i-%s)", path, errno, fd_io_strerror( errno ) ));
+  if( FD_UNLIKELY( !fgets( output, 4096, fp ) ) )
+    FD_LOG_ERR(( "error configuring network device, fgets(%s) failed (%i-%s)", path, errno, fd_io_strerror( errno ) ));
+  if( FD_UNLIKELY( feof( fp ) ) ) FD_LOG_ERR(( "error configuring network device, fgets(%s) failed (EOF)", path ));
+  if( FD_UNLIKELY( ferror( fp ) ) ) FD_LOG_ERR(( "error configuring network device, fgets(%s) failed (error)", path ));
+  if( FD_UNLIKELY( strlen( output ) == 4095 ) ) FD_LOG_ERR(( "line too long in `%s`", path ));
+  if( FD_UNLIKELY( fclose( fp ) ) )
+    FD_LOG_ERR(( "error configuring network device, fclose(%s) failed (%i-%s)", path, errno, fd_io_strerror( errno ) ));
+  output[ strlen( output ) - 1 ] = '\0';
+}
+
+static void
+init_device( const char * device ) {
+  if( FD_UNLIKELY( strlen( device ) >= IF_NAMESIZE ) ) FD_LOG_ERR(( "device name `%s` is too long", device ));
+
+  int sock = socket( AF_INET, SOCK_DGRAM, 0 );
+  if( FD_UNLIKELY( sock < 0 ) )
+    FD_LOG_ERR(( "error configuring network device, socket(AF_INET,SOCK_DGRAM,0) failed (%i-%s)",
+                 errno, fd_io_strerror( errno ) ));
+
+  struct ifreq ifr;
+  strncpy( ifr.ifr_name, device, IF_NAMESIZE-1 );
+
+  /* turn off generic-receive-offload, which is entirely incompatible with
+   * AF_XDP and QUIC
+   * It results in multiple UDP payloads being merged into a single UDP packet,
+   * with IP and UDP headers rewritten, combining the lengths and updating the
+   * checksums. QUIC short packets cannot be processed reliably in this case. */
+
+  /* command for generic-receive-offload = off */
+  struct ethtool_value gro = { .cmd = ETHTOOL_SGRO, .data = 0 };
+
+  /* attach command to ifr */
+  ifr.ifr_data = (void *)&gro;
+
+  /* log command */
+  FD_LOG_NOTICE(( "RUN: `ethtool --offload %s generic-receive-offload %u`",
+                  device, (uint)gro.data ));
+
+  /* execute command */
+  if( FD_UNLIKELY( ioctl( sock, SIOCETHTOOL, &ifr ) ) ) {
+    FD_LOG_ERR(( "configuring network device, ioctl(SIOCETHTOOL,ETHTOOL_SGRO) failed (%i-%s)",
+                 errno, fd_io_strerror( errno ) ));
+  }
+
+  if( FD_UNLIKELY( close( sock ) ) )
+    FD_LOG_ERR(( "error configuring network device, close() socket failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+}
+
+static void
+init( config_t * const config ) {
+  /* we need one channel for both TX and RX on the NIC for each QUIC
+     tile, but the interface probably defaults to one channel total */
+  if( FD_UNLIKELY( device_is_bonded( config->tiles.net.interface ) ) ) {
```

Comment:
```suggestion
  if( device_is_bonded( config->tiles.net.interface ) ) {
```
Minor code style thing - I wouldn't use branch hints on app dependent probability branches

Reply:
Michael gave me the opposite review at some point

Reply:
Can't make everyone happy 😔